E2E: Product Query: Testing atomic blocks

dinhtungdu opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->

<!-- Reference any related issues or PRs here -->

This PR adds E2E tests for atomic blocks in the context of the Product Query block, including:
- Product Image.
- Add to Cart Button.
- Product Price.

Notes:
- To make Image and Price available in the inserter, I have to remove their `parent` config.
- The Atomic blocks tests are passing in `[JavaScript E2E Tests (WP latest with Gutenberg plugin) (5)](https://github.com/woocommerce/woocommerce-blocks/actions/runs/3368914358/jobs/5587988913#logs).

<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->

### Testing

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [x] E2E tests

Running E2E test locally:
1. `npm run build`
2. `GUTENBERG_EDITOR_CONTEXT=gutenberg npm run test:e2e tests/e2e/specs/backend/product-query/atomic-blocks.test.ts`
3. See tests passing.

* [x] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

dinhtungdu committed:
wip: testing add to cart button

dinhtungdu committed:
Merge branch 'trunk' into tests/product-query-atomic-blocks

dinhtungdu committed:
test add to cart button block

dinhtungdu committed:
Merge branch 'trunk' into tests/product-query-atomic-blocks

dinhtungdu committed:
make product image available in the inserter inside product query

dinhtungdu committed:
test product image block

dinhtungdu committed:
split atomic tests to a dedicated file

dinhtungdu committed:
Merge branch 'trunk' into tests/product-query-atomic-blocks

dinhtungdu committed:
fix: make product price available in the inserter

dinhtungdu committed:
test product price block

dinhtungdu committed:
add produc rating test

sunyatasattva commented on tests/e2e/specs/backend/product-query/atomic-blocks.test.ts:
```diff
@@ -0,0 +1,127 @@
+/**
+ * Internal dependencies
+ */
+import { GUTENBERG_EDITOR_CONTEXT, describeOrSkip } from '../../../utils';
+import { shopper } from '../../../../utils';
```
In the previous meetup, we have decided to try to move away from relative paths as much as we can. Specifically, no ancenstor relative paths, because they can be hard to read and also hard to maintain. More on the reasoning in this [P2 post](https://teamkirigami.wordpress.com/2022/03/24/meetup-code-standards-and-practices-brainstorm/#relative-paths-vs-absolute-paths). 

Maybe you can create an alias in the `tsconfig` that points to the base test directory, and import from there?

dinhtungdu replied:
We have three places from importing utilities in our test at the moment:
- `tests/utils/`.
- `tests/e2e/utils.js`.
- `tests/e2e/page-utils.js`

`@woocommerce/blocks-test-utils` is actually an alias that points to `tests/utils` which contains utilities for E2E and Unit testing. We have discussed organizing test utilities [here](https://github.com/woocommerce/woocommerce-blocks/issues/5893) and [here](https://github.com/woocommerce/woocommerce-blocks/issues/6205). I think it's good to start working on https://github.com/woocommerce/woocommerce-blocks/issues/6205 after this PR. And I agree with @nielslange suggestion [here](https://github.com/woocommerce/woocommerce-blocks/issues/5893#issuecomment-1046829849):

> My thoughts are similar to @sunyatasattva's thoughts. I'd rather move all functionality to test/utils so that these functions can be used both by unit, integration and e2e tests.

sunyatasattva replied:
Alright so you are saying: let's keep the relative paths here, and then address the organization in another PR, and in that PR fix those relative paths to absolutes?

sunyatasattva commented on tests/e2e/specs/backend/product-query/atomic-blocks.test.ts:
```diff
@@ -0,0 +1,127 @@
+	'Product Query > Atomic blocks',
+	() => {
+		beforeEach( async () => {
+			await visitBlockPage( `${ block.name } Block` );
```
I suggest you add `waitForCanvas()` here in the `beforeEach`, it might not seem necessary, but since you are using `canvas()` in your test cases, believe me, it's going to eliminate some false negatives and add some safety.

dinhtungdu replied:
FYI, updated `waitForCanvas()` to support Page Editor. Gutenberg hasn't migrated Page Editor to use iframe yet.

sunyatasattva commented on tests/e2e/specs/backend/product-query/atomic-blocks.test.ts:
```diff
@@ -0,0 +1,127 @@
+		it( 'Can add the Add to Cart Button block and render it on the front end.', async () => {
+			const canvasEl = canvas();
+			await canvasEl.waitForSelector(
+				`${ block.class } .wp-block-woocommerce-product-image a img`
```
Please try to avoid as much as possible the use of [magic strings](https://deviq.com/antipatterns/magic-strings) as selectors. You can scope your selectors within your `block` object above, like was, for example, done by Luigi in `filter-products-by-attribute.test.ts` among others.

sunyatasattva commented on tests/e2e/specs/backend/product-query/atomic-blocks.test.ts:
```diff
@@ -0,0 +1,127 @@
+		} );
+
+		it( 'Can add the Product Ratings block and render it on the front end', async () => {
+			const canvasEl = canvas();
+			await canvasEl.waitForSelector(
+				`${ block.class } .wp-block-woocommerce-product-image a img`
+			);
+			const postTitleEl = await canvasEl.$(
+				`${ block.class } .wp-block-post-title`
+			);
+			await postTitleEl.click();
+			await insertBlock( 'Product Rating' );
+			await saveOrPublish();
```
All of this chunk of code is repeated across each test case. You could refactor it to a separate function, such as `insertProductQueryInnerBlock`.

sunyatasattva commented on tests/e2e/specs/backend/product-query/atomic-blocks.test.ts:
```diff
@@ -0,0 +1,127 @@
+			await canvasEl.waitForSelector(
+				`${ block.class } .wp-block-woocommerce-product-image a img`
+			);
+			const postTitleEl = await canvasEl.$(
+				`${ block.class } .wp-block-post-title`
+			);
+			await postTitleEl.click();
```
It seems that you are relying on clicking the post title to insert the block in the right place. We might migrate away from using the Post Title (and use “Product title”, for example), or certain patterns might not have that title at all. So, it seems a bit of a brittle way to describe this test, is there any other way to insert an inner block more confidently?

If not, at least the refactoring of this implementation details in a function as described in my other comment will make it easier to change it in the future. But I'd like to see if we can find a safer way to do this.

dinhtungdu replied:
I agree, I created `insertInnerBlock` as a safer way to insert inner block. It work with all parent blocks.

sunyatasattva commented on tests/e2e/specs/backend/product-query/atomic-blocks.test.ts:
```diff
@@ -0,0 +1,127 @@
+			await expect( page ).toMatchElement(
+				'.wp-block-query .woocommerce-Price-amount',
+				{
+					text: '2.99',
```
Where does this price come from?

dinhtungdu replied:
This comes from fixture data. However, I [5a3ffb7](https://github.com/woocommerce/woocommerce-blocks/pull/7482/commits/5a3ffb7df035a2b56208bf7f1382ab0db46b3def) to get the price from fixture data instead of hard coding it.

sunyatasattva reviewed (changes requested):
Thank you for tackling the tough problem of E2E tests! I left a few comments.

dinhtungdu committed:
use @woocommerce/blocks-test-utils when possbile

dinhtungdu committed:
wait for canvas first

dinhtungdu committed:
update waitForCanvas to support block editor

dinhtungdu committed:
add new utility: insertInnerBlock

dinhtungdu committed:
clarify the design decision

dinhtungdu committed:
add site editor support for insertInnerBlock

dinhtungdu committed:
update atomic blocks tests to use insertInnerBlock

dinhtungdu committed:
get price from fixture data instead of hard coding price value

dinhtungdu committed:
update expect selectors

dinhtungdu committed:
scope selectors

dinhtungdu commented:
@sunyatasattva Thank you for your review! I focused too much on getting the test done but didn't pay attention to polishing it. I addressed your comments so please take another look at this PR!

sunyatasattva commented on tests/e2e/specs/backend/product-query/atomic-blocks.test.ts:
```diff
@@ -0,0 +1,123 @@
+			await expect( canvas() ).toMatchElement(
+				block.selectors.productPrice,
+				{
+					text: fixturePrices.some( Boolean ),
```
I'm confused here… would this make it `text: true` ?

dinhtungdu replied:
My bad. I fixed this issue in 4f7d884

sunyatasattva commented on tests/utils/insert-inner-block.ts:
```diff
@@ -0,0 +1,41 @@
+		await selectBlockByName( parentBlockName );
+	}
+	const blockInserterButton = await canvas().waitForSelector(
+		'.block-editor-inserter button'
```
I'm always for refactoring these selectors somewhere else. These might change in Gutenberg and we need one easy place to make sure we can edit all the functions which uses them. It also makes them more explicit when reading the code.

We have a `SELECTORS` object inside `tests/e2e/utils.js`. If you want to create a separate file for this function, maybe you can `export` those selectors from that file? Or, actually better: if we want to make one file per function, then we should have one file for the selectors.

dinhtungdu replied:
> Or, actually better: if we want to make one file per function, then we should have one file for the selectors.

I prefer this too, implemented in [ac1222c](https://github.com/woocommerce/woocommerce-blocks/pull/7482/commits/ac1222c7c6e38bb8d61fab21c2846683207e05ea)

sunyatasattva commented on tests/e2e/specs/backend/product-query/atomic-blocks.test.ts:
```diff
@@ -0,0 +1,123 @@
+			await insertInnerBlock( 'Product Price', 'core/post-template' );
+			await expect( canvas() ).toMatchElement(
+				block.selectors.productPrice,
+				{
```
This is just a suggestion. Since `block.selectors` is used many times, you can destructure selectors in the main scope, and make the code slightly shorter. I think other test files do this.

Not a big deal if you want to leave it like this.

dinhtungdu replied:
Make sense, I destructured selectors in the main scope, and the code look cleaner.

sunyatasattva reviewed (changes requested):
Amazing work @dinhtungdu ! Those commits definitely vastly improve the quality of the code, thanks for the dedication! 🙏 

I've added some small comments. I think the main one would be the clarification of `.some`.

dinhtungdu committed:
centralize selectors for maintainability following @sunyatasattva suggestions

dinhtungdu committed:
fix random fixture price logics, destructures selectors in the main scope of the test

dinhtungdu commented:
@sunyatasattva Thanks so much for your suggestions, I applied them and this PR is ready for another review 😄

thealexandrelara reviewed (approved):
Great work @dinhtungdu! LGTM!

sunyatasattva reviewed (approved):
You are the man, @dinhtungdu ! Thank you for putting up with all my annoying nitpicks and amazing work here! Thank you so much. Let's gooo! :shipit:

dinhtungdu committed:
Merge branch 'trunk' into tests/product-query-atomic-blocks

github-actions[bot] committed:
bot: update checkstyle.xml
